maraboupy/MarabouNetworkNNet.py

```diff
@@ -218,11 +447,204 @@ def nodeTo_f(self, layer, node):
+                between layer first_layer (input layer by default)
+                and layer last_layer (output layer by default)
+
+        Args:
+            inputs: (list of floats):         Network inputs to be evaluated
+            first_layer:  (int):              the initial layer of the evaluation
+            last_layer: (int):                the last layer of the evaluation
+            normalize_inputs: (bool):         if True and first_layer==0, normalization of inputs is performed
+            normalize_outputs: (bool):        if True, normalization of output is undone
+            activate_output_layer: (bool):    if True, the last layer is activated, otherwise it is not.
+
+        Returns:
+            (list of floats): the result of the evaluation
+        """
+        num_layers = self.numLayers
+        input_size = self.inputSize
+        output_size = self.outputSize
+        biases = self.biases
+        weights = self.weights
+        mins = self.inputMinimums
+        maxes = self.inputMaximums
+        means = self.inputMeans
+        ranges = self.inputRanges
+
+        # The default output layer is the last (output) layer
+        if last_layer == -1:
+            last_layer = num_layers
+
+        inputs_norm = inputs[:]
+
+        # Prepare the inputs to the neural network
+        if normalize_inputs:
+            if (first_layer == 0):
+
+
+                for i in range(input_size):
+                    if inputs[i] < mins[i]:
+                        inputs_norm[i] = (mins[i] - means[i]) / ranges[i]
+                    elif inputs[i] > maxes[i]:
+                        inputs_norm[i] = (maxes[i] - means[i]) / ranges[i]
+                    else:
+                        inputs_norm[i] = (inputs[i] - means[i]) / ranges[i]
+            else:
+                warnings.warn('Normalization of inputs is supported for the input layer only. Request for '
+                              'normalization of inputs ignored.')
+
+        # Evaluate the neural network
+        for layer in range(first_layer, last_layer - 1):
+            inputs_norm = np.maximum(np.dot(weights[layer], inputs_norm) + biases[layer], 0)
+
+        layer = last_layer - 1
+
+        if (activate_output_layer):
+            outputs = np.maximum(np.dot(weights[layer], inputs_norm) + biases[layer], 0)
+        else:
+            outputs = np.dot(weights[layer], inputs_norm) + biases[layer]
+
+        # Undo output normalization
+        if (normalize_outputs):
+            if last_layer == num_layers:
+                output_mean = self.outputMean
+                output_range = self.outputRange
+            else:
+                output_mean = means[layer - 1]
+                output_range = ranges[layer - 1]
+
+            for i in range(output_size):
+                outputs[i] = outputs[i] * output_range + output_mean
+
+        return outputs
+
+
+    def createRandomInputsForNetwork(self):
```

Comment:
Docstring here

Reply:
@kjulian3, would you mind having another look (hopefully one last time)? Thanks!